hadoop-tools/hadoop-federation-balance/src/site/markdown/FederationBalance.md

```diff
@@ -0,0 +1,156 @@
+| -router | If `true` the command runs in router mode. The source path is taken as a mount point. It will disable write by setting the mount point readonly. Otherwise the command works in normal federation mode. The source path is taken as the full path. It will disable write by cancelling the `x` permission of the source path. The default value is `true`. |
+| -forceCloseOpen | If `true`, in DIFF_DISTCP stage it will force close all open files when there is no diff between the source path and the dst path. Otherwise the DIFF_DISTCP stage will wait until there is no open files. The default value is `false`. |
+| -map | Max number of concurrent maps to use for copy. |
+| -bandwidth | Specify bandwidth per map in MB. |
+| -moveToTrash | If `true` move the source path to trash after the job is done. Otherwise delete the source path directly. |
+
+Configuration Options
+--------------------
+
+| Configuration key              | Description                          |
+| ------------------------------ | ------------------------------------ |
+| hadoop.hdfs.procedure.work.thread.num | The worker threads number of the BalanceProcedureScheduler. Default is `10`. |
+| hadoop.hdfs.procedure.scheduler.journal.uri | The uri of the journal. |
+| federation.balance.class | The class used for federation balance. Default is `org.apache.hadoop.tools.DistCpProcedure.` |
+
+Architecture of Federation Balance
+----------------------
+
+  The components of the Federation Balance may be classified into the following
+  categories:
+
+  * Balance Procedure Scheduler
+  * DistCpFedBalance
+
+### Balance Procedure Scheduler
+
+  The Balance Procedure Scheduler implements a state machine. It's responsible
+  for scheduling a balance job, including submit, run, delay and recover.
+  The model is showed below:
+
+  ![Balance Procedure Scheduler](images/BalanceProcedureScheduler.png)
+
+  * After a job is submitted, the job is added to the pendingQueue.
+  * Worker thread takes job and run it. Journals are written to storage.
+  * If writing journal fails, the job is added to the recoverQueue for later
+    recovery. If Worker thread catches a RetryTaskException, it adds the job to
+    the delayQueue.
+  * Rooster thread takes job from delayQueue and adds it back to pendingQueue.
+  * When a scheduler starts, it will scan all the unfinished jobs from
+    journal and add them to the recoverQueue. The recover thread will recover
+    them from journal and add them back to pendingQueue.
+
+### DistCpFedBalance
+
+  DistCpFedBalance is implemented as a job of the state machine. All the distcp
+  balance logic are implemented here. A DistCpFedBalance job consists of 3
+  procedures:
+
+  * DistCpProcedure: This is the first procedure. It handles all the data copy
+    works. There are 6 stages:
+    * PRE_CHECK: Do the pre-check of the src and dst path.
+    * Init Distcp: Create a snapshot of the source path and distcp it to the
```

Comment:
Should this be two stages? One for snapshot and the other for distcp?

Reply:
In the INIT_DISTCP stage it first creates snapshot then submits distcp. In the DIFF_DISTCP stage it also needs to create snapshot and submit distcp. So I combine the snapshot and distcp.